### ISSUE TITLE
Optimise selectors like `Near` and `Contains` on `Regular` lookups

### ISSUE THREAD

**Issue by rafaqz** [rafaqz/DimensionalData.jl#708]
There is a lot of performance left in these selectors compared to just using something like `round(x / scale _ offset)` where that is possible, like with `Regular` lookups.

See:
https://github.com/rafaqz/Rasters.jl/issues/653#issuecomment-2091827421

**Comment by alex-s-gardner** [comment]
@asinghvi17, FYI this is the selector efficiency issue we were discussing

**Comment by rafaqz** [comment]
So half of this problem is that we index  into the vector to get the value at least twice - once to check that it matches the selector, and later to get it with regular indexing.

This is how how the selector abstraction works, for better or worse. Maybe it could return an object that's <: `Integer` but also has the value attached.  So we don't have to get it twice but the interface doesn't have to change and `dims2indices` still works standalone.

The other half is that `Base.searchsortedfirst` does something like `round(x / scale - offset)` anyway, but its a lot more careful about the result and fp error, so its slower. Probably we also need to be that careful.

**Comment by alex-s-gardner** [comment]
Just as an update Rasters v0.14.2 is 6x slower using `extract` than finding the index as done OP for very large arrays.

**Comment by rafaqz** [comment]
Yeah, the TwicePrecision cost means it will never match that, and we need the accuracy. But it could be just 3x slower rather than 6x
